cmd/blsync/main.go

```diff
@@ -0,0 +1,237 @@
+		kv := strings.Split(s, ":")
+		if len(kv) != 2 {
+			utils.Fatalf("Invalid custom API header entry: %s", s)
+		}
+		customHeader[strings.TrimSpace(kv[0])] = strings.TrimSpace(kv[1])
+	}
+	beaconApi := api.NewBeaconLightApi(ctx.String(utils.BeaconApiFlag.Name), customHeader)
+	committeeSyncer := api.NewCommitteeSyncer(beaconApi, chainConfig.GenesisData)
+	db := memorydb.New()
+	syncCommitteeCheckpoint := beacon.NewWeakSubjectivityCheckpoint(db, committeeSyncer, chainConfig.Checkpoint, nil)
+	if syncCommitteeCheckpoint == nil {
+		utils.Fatalf("No beacon chain checkpoint")
+	}
+	syncCommitteeTracker := beacon.NewSyncCommitteeTracker(db, chainConfig.Forks, syncCommitteeCheckpoint, ctx.Int(utils.BeaconThresholdFlag.Name), !ctx.Bool(utils.BeaconNoFilterFlag.Name), beacon.BLSVerifier{}, &mclock.System{}, func() int64 { return time.Now().UnixNano() })
+	cache, _ := lru.New(1000)
+	execSyncer := &execSyncer{
+		api:           beaconApi,
+		client:        makeRPCClient(ctx),
+		execRootCache: cache,
+	}
+	syncCommitteeTracker.SubscribeToNewHeads(execSyncer.newHead)
+	committeeSyncer.Start(syncCommitteeTracker)
+	syncCommitteeCheckpoint.TriggerFetch()
+	<-ctx.Done()
+	return nil
+}
+
+func callNewPayloadV1(client *rpc.Client, block *types.Block) (string, error) {
+	var resp cbeacon.PayloadStatusV1
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	err := client.CallContext(ctx, &resp, "engine_newPayloadV1", *cbeacon.BlockToExecutableData(block))
+	cancel()
+	return resp.Status, err
+}
+
+func callForkchoiceUpdatedV1(client *rpc.Client, headHash, finalizedHash common.Hash) (string, error) {
+	var resp cbeacon.ForkChoiceResponse
+	update := cbeacon.ForkchoiceStateV1{
+		HeadBlockHash:      headHash,
+		SafeBlockHash:      finalizedHash,
+		FinalizedBlockHash: finalizedHash,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	err := client.CallContext(ctx, &resp, "engine_forkchoiceUpdatedV1", update, nil)
+	cancel()
+	return resp.PayloadStatus.Status, err
+}
+
+type execSyncer struct {
+	api           *api.BeaconLightApi
+	client        *rpc.Client
+	execRootCache *lru.Cache // beacon block root -> execution block root
+}
+
+var statePaths = []string{
+	"[\"finalizedCheckpoint\",\"root\"]",
+	"[\"latestExecutionPayloadHeader\",\"blockHash\"]",
+}
+
+func (e *execSyncer) newHead(head beacon.Header) {
+	log.Info("Received new beacon head", "slot", head.Slot, "blockRoot", head.Hash())
+	if e.client == nil {
+		return
+	}
+	var (
+		blocks         []*types.Block
+		finalizedRoots []common.Hash
+		parentFound    bool
+	)
+	count := 16
+	for {
+		proof, err := e.api.GetStateProof(head.StateRoot, statePaths, stateProofFormat)
+		if err != nil {
+			log.Error("Error fetching state proof from beacon API", "error", err)
+			break
+		}
+		execBlockRoot := common.Hash(proof.Values[stateIndexMap[beacon.BsiExecHead]])
+		finalizedBeaconRoot := common.Hash(proof.Values[stateIndexMap[beacon.BsiFinalBlock]])
```

Comment:
It's unique per proof format. The `MultiProof` does not have a `stateIndexMap` internally so it's probably better to keep this thing external but I changed it a bit to make it more readable (now I pre-calculate and store the actual value indices instead of the map) and also added proper comments to explain the meaning of the index variables.